Build web-tree-sitter

wenkokke opened the pull request:
This pull request changes the build process of vscode-parse-tree to build web-tree-sitter, rather than depending on the published version in the NPM package registry, as it frequently lags behind the main version.

wenkokke commented:
This seems to have loading path issues?

pokey commented on .gitignore:
```diff
@@ -3,4 +3,5 @@ node_modules
 *.bin
-parsers/
\ No newline at end of file
+parsers/
+vendor/
```
Let's prefix both of these with a `/` so that we don't inadvertently ignore files vendored into `src/`, which I've had occasion to do before
```suggestion
/parsers/
/vendor/
```

pokey commented on Makefile:
```diff
@@ -1,9 +1,20 @@
+compile: \
+		$(addprefix vendor/web-tree-sitter/,$(WEB_TREE_SITTER_FILES)) \
+		$(addprefix parsers/tree-sitter-,$(addsuffix .wasm,$(LANGUAGES)))
+	tsc -p ./
```
why do you bake this one into the makefile now?  won't this change now cause us not to compile when we change any of our own source files, as they're not dependencies?

wenkokke replied:
Actually, `compile` is a phony task, so it’ll always run.

pokey commented on Makefile:
```diff
@@ -1,9 +1,20 @@
-# For generating .wasm files for parsers
-# See https://www.npmjs.com/package/web-tree-sitter
```
would be good to preserve this comment

pokey commented on Makefile:
```diff
@@ -23,4 +34,20 @@ parsers/tree-sitter-tsx.wasm: node_modules/tree-sitter-typescript/tsx/package.js
+
+# Build web-tree-sitter
+
+$(addprefix vendor/web-tree-sitter/,$(WEB_TREE_SITTER_FILES)):
```
We have no dependency on the version number, from what I can tell, so how do we trigger this one to recompile?

wenkokke replied:
Good point, I should work the version into the path

pokey reviewed (commented):
Cool stuff!  I'm not sure I understand what the workflow is supposed to be now, though.  Eg if I just clone the repo and type `yarn`, that won't work, right, because I haven't generated the `vendor/web-tree-sitter` dir yet?  But if I run `yarn compile`, that won't work either, because I haven't installed anything.  So it seems like there's a bit of a chicken and egg thing.  But I'm probably missing something

Also, a question:

- How come the `yarn.lock` keeps getting bumped?  Are you explicitly asking it to upgrade everything?

And something to fix:

- [ ] Please remove `package-lock.json`

wenkokke commented:
> How come the yarn.lock keeps getting bumped? Are you explicitly asking it to upgrade everything?

I don't know, to be honest?

wenkokke commented:
I've merged haskell-support into this pull request.

wenkokke commented:
This is the diff for Haskell support:
https://github.com/pokey/vscode-parse-tree/pull/20/commits/a5f6a921c034df0f92e7e501136d8e55fdb1735e

pokey commented:
Is this one ready for another round of review?  If not, of course feel free to re-request review when you want me to have another look

wenkokke commented:
I don’t use yarn, so I’m not sure how it differs from npm, but `npm run compile` calls `make compile`, which generates the required `vendor/web-tree-sitter/[VERSION]/` dependency.

pokey commented:
> I don’t use yarn, so I’m not sure how it differs from npm, but `npm run compile` calls `make compile`, which generates the required `vendor/web-tree-sitter/[VERSION]/` dependency.

Hmm I cloned a fresh repo on your branch, and did an `npm run compile` (or rather, a `yarn compile`, which is the same), and got the following:

```
make: *** No rule to make target `parsers/tree-sitter-agda.wasm', needed by `compile'.  Stop.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I believe that's because the dependency of `parsers/tree-sitter-agda.wasm` is `node_modules/tree-sitter-agda/package.json`, but that doesn't yet exist because nothing has been installed yet, as I haven't yet run `yarn` (equivalent to `npm install`).

If, however, I instead start by running `yarn`, I get the following error:

```bash
➜ yarn                 
yarn install v1.22.17
warning ../../package.json: No license field
[1/5] 🔍  Validating package.json...
warning parse-tree@0.14.0: The engine "vscode" appears to be invalid.
[2/5] 🔍  Resolving packages...
error Package "web-tree-sitter" refers to a non-existing file '"/Users/pokey/src/vscode-parse-tree-test/vendor/web-tree-sitter/v0.20.4"'.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

Because I haven't yet run `yarn compile`.  That's what I meant by a chicken-and-egg thing.  You need to compile to create the proper dependencies, but then you need the proper dependencies to compile

I think we probably need to split the makefile's `compile` rule into two separate rules: one to build the web-tree-sitter package, and one to compile.  You would run the first rule, then run `yarn`, then run the second rule.  Make sense?

wenkokke commented:
Changes:

- Build web-tree-sitter in the preinstall script, rather than the compile script;
- Added a patch to web-tree-sitter to bump the version number;
- Add the vendor and .make-work directories to the list of excluded paths;
- Removed the old version of web-tree-sitter from yarn.lock.

pokey reviewed (commented):
Ok so this all looks great!  +70 for persistence 😅.  I updated the contributing instructions; they haven't really changed it seems but I find it works better if I run `yarn compile` so that emscripten ends up on the path.  A couple things left:

- [ ] Would be good to update the docs with instructions on how to upgrade `web-tree-sitter`, ie: that patch file, etc
- [ ] I'm still getting those failures on the cursorless html test cases after installing this branch locally.  If you're unsure how to debug the cursorless side of things I'd be happy to pair

```
stack trace: TypeError: Cannot read property 'apply' of undefined
	at [/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/node_modules/web-tree-sitter/tree-sitter.js:1:10374]()
	at _ZN3TagC2E7TagTypeRKNSt3__212basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE (<anonymous>:wasm-function[141]:0xad34)
	at _ZN3Tag8for_nameERKNSt3__212basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEE (<anonymous>:wasm-function[125]:0xaa8e)
	at tree_sitter_html_external_scanner_scan (<anonymous>:wasm-function[25]:0x9d5e)
	at ts_parser_parse_wasm (<anonymous>:wasm-function[214]:0x24857)
	at Parser.parse ([/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/node_modules/web-tree-sitter/tree-sitter.js:1:35040]())
	at [/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/out/extension.js:82:43]()
	at Generator.next (<anonymous>)
	at fulfilled ([/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/out/extension.js:5:58]())
	at runMicrotasks (<anonymous>)
	at processTicksAndRejections (internal[/process/task_queues.js:93:5]())
```

wenkokke commented on tree-sitter.patch:
```diff
@@ -0,0 +1,12 @@
+ {
+   "name": "web-tree-sitter",
+-  "version": "0.20.3",
++  "version": "0.20.4",
```
It’s just to fix the version number. The actual tag used is v0.20.4, but they didn’t increment the web-tree-sitter version. The patch will always depend on the version we compile against, so it’ll have to change when we bump the web-tree-sutter version, yes.

wenkokke replied:
> Should prob add a quick bullet-point section to the README indicating how to bump web-tree-sitter

Fair, yes.

wenkokke commented:
> Ok so this all looks great! +70 for persistence 😅. I updated the contributing instructions; they haven't really changed it seems but I find it works better if I run `yarn compile` so that emscripten ends up on the path. A couple things left:
> 
> * [ ]  Would be good to update the docs with instructions on how to upgrade `web-tree-sitter`, ie: that patch file, etc
> * [ ]  I'm still getting those failures on the cursorless html test cases after installing this branch locally.  If you're unsure how to debug the cursorless side of things I'd be happy to pair
> 
> ```
> stack trace: TypeError: Cannot read property 'apply' of undefined
> 	at [/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/node_modules/web-tree-sitter/tree-sitter.js:1:10374]()
> 	at _ZN3TagC2E7TagTypeRKNSt3__212basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE (<anonymous>:wasm-function[141]:0xad34)
> 	at _ZN3Tag8for_nameERKNSt3__212basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEE (<anonymous>:wasm-function[125]:0xaa8e)
> 	at tree_sitter_html_external_scanner_scan (<anonymous>:wasm-function[25]:0x9d5e)
> 	at ts_parser_parse_wasm (<anonymous>:wasm-function[214]:0x24857)
> 	at Parser.parse ([/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/node_modules/web-tree-sitter/tree-sitter.js:1:35040]())
> 	at [/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/out/extension.js:82:43]()
> 	at Generator.next (<anonymous>)
> 	at fulfilled ([/Users/pokey/.vscode/extensions/pokey.parse-tree-0.14.0/out/extension.js:5:58]())
> 	at runMicrotasks (<anonymous>)
> 	at processTicksAndRejections (internal[/process/task_queues.js:93:5]())
> ```

How do I reproduce this HTML error?

pokey commented:
You should be able to do a local package and install of this extension, then clone cursorless and run unit tests locally

pokey commented:
Ok as discussed this one now seems to be working for me with cursorless locally, so just docs on how to upgrade web-tree-sitter and I think we're good to go!

wenkokke commented:
I have added a short section documenting how to update `web-tree-sitter`, and I've added the version number to the patch file so that we can store some historical patches if we need to, and to make the upgrade process easier, e.g., no need for creating empty patches or for changing the patching code.

ParetoOptimalDev commented:
@wenkokke Are you also working on support for Haskell in cursorless-dev/cursorless-vscode?

If not I might look in too starting since cursorless could make my work week a lot easier :smile:

wenkokke commented:
Yes, that’s next on my calendar, but if you’d like to have a chat I’d be happy to!
